### DISCUSSION TITLE
Highlight which ambassadors are on stream

### DISCUSSION THREAD

**Discussion by pjeweb** [alveusgg/extension#22] [Ideas]
# Idea

I would like the extension to somehow highlight not only one Ambassador (popup), but also _all_ ambassadors that are currently on stream or on cam (pasture = highlight Stompy, Donks, Winnie; crows= highlight Abbot and Coconut; etc.)

## How to sync with stream:

  1) Get channel id (See #33)
  2) Fetch the title on load using the Helix API [1], parse any known commands in the title (e.g. !nilla and/or !pasture), update the UI
  3) <strike>Listen for channel Updates using the websocket eventsub [2] [3] and parse it</strike> => This is not possible from within the extension. This would require a backend service that could push it using the extension pubsub
  4) Listen for mod/bot messages in the Chat, then either fetch the title again and parse it or get the commands from the chat message
  5) and/or poll the Helix API every few minutes for the current title

### Alternatives:

  - Add a new mod command or command modifier that does not pop up the ambassador but only highlights it in the list `!highlight winnie stompy jalapeno serano` and `!hlreset`
  - <strike>Some crazy OCR to detect what ambassadors are on stream like windowtowildlife?!</strike>

## UI

  - The most basic UI would be differently coloured outline around highlighted ambassadors
  - Another option would be to keep highlighted ambassadors at the top of the list (maybe even with a separator)

-----------------------

1) https://dev.twitch.tv/docs/api/reference/#get-channel-information
2) EventSub using WS: https://dev.twitch.tv/docs/eventsub/handling-websocket-events/
3) EventSub channel update: https://dev.twitch.tv/docs/eventsub/eventsub-subscription-types/#channelupdate

**Comment by abdullahmorrison** [comment]
This should be done. Not super high priority tho.
- I don't think a bot needs to send a message chat, just change the extension.
- We don't need !hlreset, it can just auto reset after a few secs. 
- The outline can be green colored.

I don't think persistence is necessary. It's not a big deal if a user joins later. If they miss it, they miss it.

OCR: if it can be done with a high success rate, sure.

**Reply by pjeweb** [reply]
For keeping/educational streams I totally agree that your great solution of listening for chat commands in client would also work great for highlighting multiple ambassadors, but I think having, them somewhat more persistent would be great for the 24/7 live cams where a scene with a few ambassadors may be live for a long time. So when a random person browsing twitch comes across the live stream they see the "faces" of the ambassadors on stream to click and learn more.

**Reply by abdullahmorrison** [reply]
I'm open to it.

**Reply by MattIPv4** [reply]
In terms of persistence, I wonder if you could work with Space to have the scene management for the live cams include what scene is active in the title, so you could then parse the stream title to tell what should be highlighted? For backpack streams etc., you could just rely on the command and not worry about persistence for late joining.

**Reply by pjeweb** [reply]
Yes, thats a great solution. Maybe we could add group/cam commands to make it a bit shorter (`!pasture` `!crows` `!parrots` `!chickens`).
